src/tss2-tcti/tcti-tbs.c

```diff
@@ -0,0 +1,286 @@
+    }
+    if (*response_size < tcti_tbs->resSize) {
+        LOG_WARNING ("Caller provided buffer that is not large enough to "
+                  "hold the response buffer.");
+        rc = TSS2_TCTI_RC_BAD_VALUE;
+        goto out;
+    }
+    if (tcti_tbs->resSize == 0 || tcti_tbs->resSize > TBS_RESULT_MAX_BUFFER_SIZE) {
+        LOG_WARNING ("The size of the response buffer is invalid, resSize=%d", tcti_tbs->resSize);
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+    memcpy (response_buffer, tcti_tbs->resultBuffer, tcti_tbs->resSize);
+    LOGBLOB_DEBUG (response_buffer, tcti_tbs->resSize, "Response Received");
+    rc = header_unmarshal (response_buffer, &tcti_common->header);
+    if (rc != TSS2_RC_SUCCESS) {
+        goto out;
+    }
+    /*
+     * Executing code beyond this point transitions the state machine to
+     * TRANSMIT. Another call to this function will not be possible until
+     * another command is sent to the TPM.
+     */
+out:
+    tcti_common->state = TCTI_STATE_TRANSMIT;
+    tcti_tbs->resSize = 0;
+    return rc;
+}
+
+void
+tcti_tbs_finalize (
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs = tcti_tbs_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_tbs_down_cast (tcti_tbs);
+    TBS_RESULT tbs_rc;
+
+    if (tcti_tbs == NULL) {
+        return;
+    }
+
+    free (tcti_tbs->resultBuffer);
+
+    tbs_rc = Tbsip_Context_Close (tcti_tbs->hContext);
+    if (tbs_rc != TBS_SUCCESS) {
+        LOG_WARNING ("Failed to close context with TBS error: 0x%x", tbs_rc);
+    }
+
+    tcti_common->state = TCTI_STATE_FINAL;
+}
+
+TSS2_RC
+tcti_tbs_cancel (
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    TBS_RESULT tbs_rc;
+    TSS2_RC rc = TSS2_RC_SUCCESS;
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs = tcti_tbs_context_cast (tctiContext);
+
+    tbs_rc = Tbsip_Cancel_Commands (tcti_tbs->hContext);
+    if (tbs_rc != TBS_SUCCESS) {
+        LOG_WARNING ("Failed to cancel commands with TBS error: 0x%x", tbs_rc);
+        rc = TSS2_TCTI_RC_GENERAL_FAILURE;
+    }
+
+    return rc;
+}
+
+TSS2_RC
+tcti_tbs_get_poll_handles (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    TSS2_TCTI_POLL_HANDLE *handles,
+    size_t *num_handles)
+{
+    /* TBS doesn't support polling. */
+    (void)(tctiContext);
+    (void)(handles);
+    (void)(num_handles);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+TSS2_RC
+tcti_tbs_set_locality (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    uint8_t locality)
+{
+    /*
+     * TBS currently only supports locality 0
+     */
+    (void)(tctiContext);
+    (void)(locality);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+TSS2_RC
+Tss2_Tcti_Tbs_Init (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t *size,
+    const char *conf)
+{
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs;
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common;
+    TBS_RESULT tbs_rc;
+    TBS_CONTEXT_PARAMS2 params;
+    TPM_DEVICE_INFO info;
+
+    if (tctiContext == NULL && size == NULL) {
```

Comment:
Could this maybe be re-written? It seems a little difficult to follow where the `else` case might end up. Perhaps something like this?
```
if (tctiContext == NULL) {
    if(size == NULL) {
        return TSS2_TCTI_RC_BAD_VALUE;
    }
    *size = sizeof (TSS2_TCTI_TBS_CONTEXT);
    return TSS2_RC_SUCCESS;
}
```
Then you'd only be checking `tctiContext` once?

Reply:
Fixed. That looks a lot better, thanks.